New skew integration algorithm for single crystal Bragg peaks

RichardWaiteSTFC opened the pull request:
**Description of work.**

A new integration method that minimises the skew of background points to identify pixels and bins contributing to a peak.

![image](https://user-images.githubusercontent.com/55979119/179549978-20d1ce68-4e6c-45bb-8b99-3e25aa01f9f6.png)


**To test:**
1. Load data and add peaks
```
Load(Filename=r'C:\mantid\build\ExternalData\Testing\Data\SystemTest\SXD23767.raw', OutputWorkspace='SXD23767')
CreatePeaksWorkspace(InstrumentWorkspace='SXD23767', NumberOfPeaks=0, OutputWorkspace='SingleCrystalPeakTable')
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=3271.0941922690058, DetectorID=32615, Height=48579, BinCount=1748.6544047163125)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=3806.3122671465112, DetectorID=30017, Height=28792, BinCount=2623.6364687743135)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=5009.848706357172, DetectorID=42513, Height=8312, BinCount=683.91325512745721)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=2485.820351956645, DetectorID=39170, Height=594, BinCount=47.772653645954279)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=2613.5255156927979, DetectorID=40740, Height=755, BinCount=73.380982729251286)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=2328.3963413700967, DetectorID=40905, Height=509, BinCount=64.125675193134072)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=1703.2033622296876, DetectorID=28764, Height=1512, BinCount=155.7961158603562)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=2449.3905245093624, DetectorID=41068, Height=2742, BinCount=352.01543879310225)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=2303.3581784155558, DetectorID=40121, Height=265, BinCount=0.00021976268225073342)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=2038.5978951201726, DetectorID=36910, Height=517, BinCount=71.174071530741671)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=3787.5542525102373, DetectorID=42749, Height=14856, BinCount=1941.4412168797021)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=885.94412995172002, DetectorID=14284, Height=216, BinCount=7.1705279215654789)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD23767', TOF=1817.2221035426605, DetectorID=36104, Height=118, BinCount=11.982004832328599)    
```

2. Call integration algorithm
```
IntegratePeaksSkew(InputWorkspace='SXD23767', PeaksWorkspace='SingleCrystalPeakTable', 
    OutputWorkspace='out', OutputFile=r"C:\Users\xhg73778\Desktop\out_v9.pdf",
    UseNearestPeak=False, UpdatePeakPosition=False, IntegrateIfOnEdge=False, 
    BackScatteringTOFResolution=0.04, ThetaWidth=0.02, )
```
Check output file and confirm peaks with masks on edge are not integrated.

3. Set `IntegrateIfOnEdge =True` - check peaks with mask on edge are now integrated (unless failing other mask criteria)
4. Set `UseNEarestPeak = True` - check that peaks with center outside mask are integrated
5. Set `UpdatePeakPosition=True`  - check peak position is moved to pixel with max. intensity.
6. Play with `NColsMax` and `NRowsMax` to exclude peaks above a certain size
7. Play with `NVacanciesMax` and `NPixPerVacancyMin` to exclude peaks with vacancies (non-peak pixels surrounded by peak pixels)


Fixes #33725
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

RichardWaiteSTFC committed:
Add new skew integration method (WIP)

RichardWaiteSTFC committed:
Add new skew integration method

RichardWaiteSTFC committed:
Add option for constant TOF width

Disable TOF resolution params if tof window > 0

RichardWaiteSTFC committed:
Save plots to pdf and make ouptut file optional

RichardWaiteSTFC committed:
Supply dTOF/TOF instead of constant dTOF when no resolution params

Also clarified parameter doc strings

RichardWaiteSTFC committed:
Add array converter for WISH-like instruments

RichardWaiteSTFC committed:
Correct typo in WISH instrument definition

RichardWaiteSTFC committed:
Set upper bound on validators for fractional TOF width

RichardWaiteSTFC committed:
Add peak mask status to printed plots

RichardWaiteSTFC committed:
Add option to move peak position

RichardWaiteSTFC committed:
USe child algorithms instead of mantid.simpleapi

RichardWaiteSTFC committed:
Plot updated peak TOF

RichardWaiteSTFC committed:
Catch IOError on opening plot file with try statement

RichardWaiteSTFC committed:
import matplotlib inside exec to fix FindPeaksAutomatic test

RichardWaiteSTFC committed:
Fix bugs for small TOF slice and no peak in 1D and nearest peak find

RichardWaiteSTFC committed:
Fix array converter for rectangular banks with idfillbyfirst_y=False

RichardWaiteSTFC committed:
Add basic tests for IntegratePeaksSkew

RichardWaiteSTFC committed:
Make setup a class method and add output file test

RichardWaiteSTFC committed:
Add tests for peak mask validation

RichardWaiteSTFC committed:
Improve plot formatting

RichardWaiteSTFC committed:
Fix bug in skew loop that added extra point to peak peak pixels

RichardWaiteSTFC committed:
Clarify parameter doc string

RichardWaiteSTFC committed:
Fix matplotlib deprecation error on RHEL

RichardWaiteSTFC committed:
Fix bug finding num rows and cols in mask (needed swapping)

RichardWaiteSTFC committed:
Add option to optimise mask based on optimal tof window found

RichardWaiteSTFC committed:
Fix bug where could supply NPixPerVacancyMin < 1

RichardWaiteSTFC committed:
Fix bug on integration status if optimised mask touches detector edge

RichardWaiteSTFC committed:
Add algorithm documentation page

RichardWaiteSTFC committed:
Add parameter to specify different num of rows and cols in 2D window

And added test

RichardWaiteSTFC committed:
Add params to extend the num. pixels of the detector edge and tests

Also added validation.
Note that for ComponentArray type detectors the number of edge pixels for columns must be one because
unless a window overlaps the edge of a detector it is not possible discriminate the two possible scenarios:
1. The window exceeds the detector with no adjacent bank
2. The window contains pixels from two banks
Because the code does not search adjacent banks for scenario 2.

RichardWaiteSTFC committed:
Add input validation for peak mask dimension limits given window size

And add test, also refactor edge pixel test

RichardWaiteSTFC committed:
Support ncols_edge > 1 for CompnentArray detector banks

Removed corresponding input validation

RichardWaiteSTFC committed:
Add tests for compnent array detectors

RichardWaiteSTFC committed:
Merge branch 'main' into 33725_new_skew_integration_algorithm

RichardWaiteSTFC committed:
Add release notes for new algorithm

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+                ax[1].autoscale_view()
+                pdf.savefig(fig, rasterized=False)
+                [subax.clear() for subax in ax]  # clear axes for next figure rather than make new one (quicker)
+                cbar.remove()
```
Similar comment to previous one - perhaps you could pull out L642-683 into a plot_peak function so that exec isn't quite so large?

RichardWaiteSTFC replied:
Good suggestion thanks!

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+                    L = (np.sin(th) ** 2) / (wl ** 4)  # at updated peak pos if applicable
+                    intens = L * np.sum(ypk[ixlo_opt:ixhi_opt])
+                    sig = L * np.sqrt(np.sum(epk_sq[ixlo_opt:ixhi_opt]))
+                    integrated = True
```
The exec function is quite large at the moment. Perhaps you could pull out L589-637 into a function? If there's lots of variables need using after this point then not a "must do" but worth checking if this is possible

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+        return next_col_prefix, next_col_str, isLHS, isRHS
+
+    def _get_detid_array_comp_assembly(self, bank, detid, row, col, drows, dcols, nrows_edge, ncols_edge=1):
+        ndet_per_spec = len(self.ws.getSpectrum(self.ws.getIndicesFromDetectorIDs([detid])[0]).getDetectorIDs())
```
Please can you separate this line out a bit. If I run the test script on the PR with InputWorkspace set to a workspace other than SXD23767 (eg one from another instrument) I get an "index out of range" error from this line. This happens because self.ws.getIndicesFromDetectorIDs([detid]) returns an empty list. In that case I think this function should log an error and then the for loop in exec should carry on with the next peak

RichardWaiteSTFC replied:
So passing the error back up the function chain is a bit messy, so instead I have added additional validation and checks at a higher level: in the input validation I check the peak and data workspaces have the same instrument and in the exec loop over all peaks I check the detector of each peak has a valid component index (i.e. exists in the instrument) and that the detector is not a monitor or masked. I think this will fix the error you found.

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+        detids = pk_ws.column('DetID')
+        bank_names = pk_ws.column('BankName')
+        tofs = pk_ws.column('TOF')
+        if plot_filename:
```
Perhaps this should also check that pk_ws has at least one peak so that an empty pdf is never created? One of my other comments is about adding some validation in the peak loop so perhaps this should only be created if the integration succeeds for at least one peak in pk_ws which would mean creating later on after the 1st peak is successfully integrated

RichardWaiteSTFC replied:
I have added a check that the peak workspace has at least one peak in the input validation - I still want to plot cases where the integration fails because the colorfill plot shows why it failed (i.e. if the mask wasn't valid), which may be informative to the user.

DannyHindson replied:
I think if the detid check that you've added results in a "continue" on L640 then you could end up with an empty pdf? I agree that the extra validation closes off one route to getting an empty pdf so see how much time you have later this week for addressing this - can skip if run out of time

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+                img = ax[0].imshow(image_data, norm=LogNorm(vmax=image_data[peak_mask].mean()))
+                ax[0].plot(*np.where(peak_mask.T), 'xw')
+                ax[0].plot(icol, irow, 'or')
+                title_str = f"{ipk} ({str(pk.getIntHKL())[1:-1]}) " \
```
Might be worth putting the DetID from the peaks workspace in here rather than (or in addition to) the peak index?

RichardWaiteSTFC replied:
The DetID doesn't necessarily uniquely identify a peak (e.g 002 and 004 will be in the same detector) - but there's no harm in adding it as well!

DannyHindson replied:
Didn't see a change for this but your observation about it not uniquely identifying the peak means it's not crucial

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+        :return ispec: spectrum index corresponding to detid of peak
+        :return edges: bool mask True for pixels on the edge of a detector bank/panel
+        """
+        bank = self.inst.getComponentByName(bank_name)
```
I think it's worth checking bank isn't `None` here (or inside get_detid_array). If the peaks workspace doesn't line up with the input workspace this might happen. Again this comes from running the test script on the PR with an input workspace instead of SXD23767

RichardWaiteSTFC replied:
I will also add a check in the input validation that the instrument of the peak workspace is the same as the raw workspace!

DannyHindson replied:
OK that sounds good. Hadn't realised the peaks workspaces had an instrument

RichardWaiteSTFC replied:
As in a previous comment I have added input validation to make sure the pea and data workspace have the same instrument - if that's the case then the bank name must correspond to a instrument component.

DannyHindson reviewed (changes requested):
I've reviewed everything apart from the tests so thought I'd share my comments so far before it gets too late in the day. I might come back to this later
Looking very good - good job!

DannyHindson commented on docs/source/algorithms/IntegratePeaksSkew-v1.rst:
```diff
@@ -0,0 +1,160 @@
+
+7. Increase the TOF window until the ratio of :math:`I/\sigma` reaches a maximum.
+
+8. If ``OptimiseMask=True`` then steps 4-7 will be repeated using the optimal TOF window to find the detectors
```
Is it steps 3-7 that get repeated? Looks like `find_peak_mask` gets called again which includes the skew minimisation

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+    return np.sum(signal) / np.sqrt(np.sum(error_sq))
+
+
+def optimise_window_intens_over_sig(signal, error_sq, ilo, ihi, ntol=8):
```
is the window here a tof window as opposed to a detector window? If yes then would help clarify if you add "tof" in the name

RichardWaiteSTFC replied:
Yep, I will make the function name something like
`optimise_tof_window_intens_over_sig`
as you suggest!

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+
+
+def find_peak_mask(signal, ixlo, ixhi, irow, icol, use_nearest):
+    _, ipeak2D = find_bg_pts_seed_skew(signal[:, :, ixlo:ixhi].sum(axis=2).flatten())
```
Please can you add a comment indicating what dimension you're calculating the skew over here. I think you're calculating it over detectors (which is different to the other call to `find_bg_pts_seed_skew` where it's being calculated over TOF bins). It's not always totally obvious what the 3 dimensions of the signal array are so the comment would help clarify this

RichardWaiteSTFC replied:
The function `find_bg_pts_seed_skew` works on any 1D array - it is used both to search for background points in the tof window and the detector window (the latter array is flattened). In this case it is used in the detector window (after summming over a TOF range). I will add a comment and maybe change the function name to `find_2D_peak_mask` or similar?

DannyHindson replied:
OK great. I think the function name is OK (was clear to me what the peak mask was). Just lost my bearings for a while about what was going on in `find_bg_pts_seed_skew` until I realised it was general and worked on any 1D array

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+            detLHS = self.inst.getComponentByName(
+                delim.join([bank_name, f'{col_prefix}{1:0{len(col_str)}d}']))
+            detRHS = self.inst.getComponentByName(
+                delim.join([bank_name, f'{col_prefix}{ncols:0{len(col_str)}d}']))
```
I think there's an assumption in here that the "module" names end in a number. Perhaps worth documenting this somewhere (unless it's easy to generalise to the case where the module name doesn't end in a number).

I spotted this when running on a daft input workspace for another instrument (POLARIS) where the module names are things like "bank1module", "bank2module" and in this case the detector name passed into `getComponentByName` isn't constructed properly

RichardWaiteSTFC replied:
I have assumed the component names (for rows and columns) end with an integer as this is the assumption made in the code which populates the row and column in peak object/workspace (so a def-facto standard for single-crystal diffractometers with component array detector banks? ). 
https://github.com/mantidproject/mantid/blob/bc222196e4303a5ce799455408e412eb4c0b04a2/Framework/DataObjects/src/Peak.cpp#L229-L241
I agree that I should state this in the docs though (also the assumption that the indexing starts from 1 not 0 - i.e. tube001 is the first tube. And probably also throw a warning if no component is found 9as per a previous comment).

RichardWaiteSTFC replied:
Perhaps I should strip the component name of any trailing non-digit characters?

RichardWaiteSTFC replied:
Just checked on POLARIS (component array banks) and making a `SingleCrystalPeakTable` using pick tool produces peaks with column = -1 (which wouldn't work with this approach anyway). So I will just document the assumptions made about the instrument component names.

DannyHindson reviewed (changes requested):
Just had another look through including the tests. Didn't see any problems with the tests, all look good. Have added one or two other comments on the main algorithm code.
Appreciate it's a bit late in the day and you're off next week so no problem to pick up when you're back

RichardWaiteSTFC committed:
Add validation for peaks workspace (number peaks, same instrument)

RichardWaiteSTFC committed:
Check if detector is monitor or masked

RichardWaiteSTFC committed:
Merge branch 'main' into 33725_new_skew_integration_algorithm

RichardWaiteSTFC committed:
Handle invalid detctor ID for peaks predicted outside detector

RichardWaiteSTFC committed:
Put plotting in separate function (outside exec)

RichardWaiteSTFC committed:
Improve maximising I/sig of 1D window

Subtract a bg from 5 points at tend oif the window (helps capture tail of peak) - force min value of bg to be 0 (as apply this post background subtraction of the data) and only replace bg value if lower than previous iteration (this is good for when there are overlapping peaks).

RichardWaiteSTFC committed:
Rename a function and params NCols and NRows (to DCols and DRows)

Other minor chanegs to doc strings etc. as requested in review

RichardWaiteSTFC committed:
Add estimation of TOF resolution parameters

RichardWaiteSTFC committed:
Update documentation and add section on assumptions about instrument

RichardWaiteSTFC commented on Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:
```diff
@@ -0,0 +1,714 @@
+                             doc="Half number of rows in the window on detector.")
+        self.declareProperty(name="NCols", defaultValue=8, direction=Direction.Input,
+                             validator=IntBoundedValidator(lower=1),
+                             doc="Half number of columns in the window on detector.")
```
I have renamed these parameters to e.g. `DRows` - to distinguish them from other parameters like `NRowMax` and have clarified the doc strings and documentation.

DannyHindson replied:
The description is much more clear now thanks. What does the new "D" prefix on the variable names indicate?

RichardWaiteSTFC replied:
In my head `D` stands for delta - so I tend to use it when I have a value +/- delta - which I admit is a bit arbitrary )it could also refer to the total width i suppose) and not obvious to users...I guess I didn't want to have `N` prefix as other parameters with `N` refer to the total number of rows/columns in e.g. the mask. I'm open to any variable name suggestions - I'm not completely happy with `DRows` etc!

DannyHindson replied:
OK I see. I would normally associate a delta with a floating point quantity so perhaps that doesn't feel quite right here. How about just adding HW for half width in somewhere eg NRowsHW and NColsHW?

RichardWaiteSTFC commented:
Since the last review I have added a couple of things (in addition to the comments):

1. More robust algorithm to find maximise I/sigma of TOF window (that uses background shell)
2. Provide the user with an estimate of appropriate TOF resolution parameters given the optimal TOF window found.

DannyHindson reviewed (changes requested):
I've tested this again and have added a few comments - most of them are caveated by "if you've got time". The new validation on the instrument being the same in the input and peak workspaces has removed the edge\error cases I've found so thanks for that. None of these things are worth missing the end of week cutoff for the release so let's see how your week goes.

RichardWaiteSTFC commented:
> I think if the detid check that you've added results in a "continue" on L640 then you could end up with an empty pdf? I agree that the extra validation closes off one route to getting an empty pdf so see how much time you have later this week for addressing this - can skip if run out of time

Indeed, this is because atm the plotting is inside the loop over peaks  where this check is done. There are two obvious things that can be done (can actually do both):

1. Put the check outside the exec - e.g. put a loop over all peaks in the validation to check that there is at least one peak with a valid detid (and break loop once found one).
2. Store the results of the integration in an object (i.e. extend the peak_data object you suggested) and put the figure creation and plotting in a separate loop after the integration.

I think (1) is quick and worth doing to avoid an empty pdf, I was considering option (2) anyway to speed up the plotting (e.g. can put ~3-5 peaks on a pdf page to avoid the number of separate write events (which I expect will help speed it up?).

DannyHindson commented:
> I think (1) is quick and worth doing to avoid an empty pdf, I was considering option (2) anyway to speed up the plotting (e.g. can put ~3-5 peaks on a pdf page to avoid the number of separate write events (which I expect will help speed it up?).

Thanks for having a look. Doing 1) only or both 1) and 2) is fine with me. Just see what you've got time for.

RichardWaiteSTFC committed:
Change param to NRows/NCols - now total num of components in window

Add validation to force values to be odd (so there is a central pixel) and clarify the definition of NRow in the case of n-1 detector mapping.

RichardWaiteSTFC committed:
Do not print estimated resolution parameters if not physical

RichardWaiteSTFC committed:
Move code form exec to methods of PeakData object

RichardWaiteSTFC committed:
Move plotting to separate loop afdter integration loop

Avoids edge case where could save an empty pdf

DannyHindson reviewed (approved):
Thanks for the changes. All looks good now. This is a very nice addition to the single crystal functionality.
